Build no amp in roadm and add roadm restrictions on amps

EstherLerouzic opened the pull request:
This PR adds the amp restriction in Roadms (booster and preamp). 
to use it with xls input file
-user may enter (valid) restrictions in eqpt_config.json and all roadms will have same restriction
-user may add per Roadm restriction in Nodes sheet (column 9 and 10 for booster and preamp, respectively). this will override eqpt_config general input for this particular node
to use it with json:
- user may enter (valid) restrictions in eqpt_config.json and all roadms will have same restriction
- user must add a valid restriction field in the json roadm element. this will override general input
Note that restrictions apply to the entire roadm, and not per degree.

a special case without booster is enabled via the xls  eqpt sheet: if a Roadm amp is marked fused, this will add a Fused element instead of an amp, enabling special design of roadms without booster. This can be applied per degree

This PR refers to #211 discussion. @jeanluc-auge, @ojnas , @jktjkt could you please have a look and tell me how to improve ? Thanks !

ojnas commented on gnpy/core/network.py:
```diff
@@ -349,6 +360,8 @@ def set_egress_amplifier(network, roadm, equipment, pref_total_db):
         #go through all nodes in the OMS (loop until next Roadm instance)
             if isinstance(node, Edfa):
                 node_loss = span_loss(network, prev_node)
+                # print('dp',prev_dp)
```
I think we already use [`logging`](https://docs.python.org/3/library/logging.html) in some places?

EstherLerouzic replied:
Indeed we already have logging in place, and this one was just a leftover from a intermediate commit. I ll check that I have removed all of them

EstherLerouzic commented on gnpy/core/elements.py:
```diff
@@ -126,14 +126,17 @@ def __init__(self, *args, params, **kwargs):
+                'params'    : {
+                    'target_pch_out_db' : self.effective_pch_out_db,
+                    'restrictions'      : self.restrictions},
+                    'metadata'      : {
```
OK! I correct !
thankx

EstherLerouzic commented on gnpy/core/network.py:
```diff
@@ -42,6 +42,21 @@ def save_network(filename, network):
+        if key in d1:
+            if isinstance(d1[key],dict):
+                c[key] = upd(c[key],d2[key])
+        else:c[key] = d2[key]
```
OK, I correct!

EstherLerouzic commented on gnpy/core/network.py:
```diff
@@ -42,6 +42,21 @@ def save_network(filename, network):
 
+
+
+def upd(d1,d2):
```
OK, I have renamed the function

EstherLerouzic commented on gnpy/core/network.py:
```diff
@@ -124,7 +119,15 @@ def select_edfa(raman_allowed, gain_target, power_target, equipment, uid, restri
+    edfa_dict = equipment['Edfa'].copy()
+    # for roadm restriction only : create a dict including not allowed for design amps
+    # because main use case is to have specific radm amp which are not allowed for ILA
+    # with the auto design
```
OK (with name in restriction :)

EstherLerouzic commented on gnpy/core/network.py:
```diff
@@ -386,20 +379,20 @@ def set_egress_amplifier(network, roadm, equipment, pref_total_db):
 
                 # implementation of restrictions on roadm boosters
                 if isinstance(prev_node,Roadm):
+                    if '' in prev_node.restrictions['booster_variety_list']: prev_node.restrictions['booster_variety_list'].remove('')
```
Ok, I added a function that removes el and does nothing (no Error) if el is not in the list.

EstherLerouzic commented on tests/data/eqpt_config.json:
```diff
@@ -75,8 +84,8 @@
             "target_pch_out_db": -20,
             "add_drop_osnr": 38,
             "restrictions": {
-                            "preamp_variety_list":["low_gain_preamp", "high_gain_preamp"],
```
"low_gain_preamp" and "high_gain_preamp" where probably given as examples names but these amp where not defined in the edfa list. Since the feature was not coded, there was not effect or impact.  Now that I have coded and activated the feature there should be valid names here. But in this case this will change many tests. So I prefer to leave it empty for now, and to create a specific test for restrictions later on.

jktjkt replied:
Ah, right, thanks. It was not implemented before :).

EstherLerouzic commented:
Please do not merge yet: I am still working on the associated tests

jktjkt commented on gnpy/core/utils.py:
```diff
@@ -199,3 +199,42 @@ def rrc(ffs, baud_rate, alpha):
+
+    >>> d3 = {'params': {'restrictions': {'preamp_variety_list': ['foo'], 'booster_variety_list': ['bar']}}}
+    >>> merge_dicts_recursively(d1, d3)
+    {'params': {'restrictions': {'preamp_variety_list': [], 'booster_variety_list': []}}}
```
@EstherLerouzic , can you please confirm that this is a correct behavior? I don't know whether it would make more sense to merge the leaf lists or not. Either way, this test just documents the current behavior.

EstherLerouzic replied:
Thanks for the commits !
This is the correct behaviour.  Restriction read from node overrides restriction from eqpt_config.json if they exist, so we should not merge preamp and booster lists.
so maybe the naming is not correct, it is more an update

jktjkt replied:
- Please feel free to squash my changes to the commits which introduce that code. Or not, I do not really have a preference here.
- Also feel free to rename this function if you feel like doing that. Perhaps something like `merge_amplifier_restrictions`?

EstherLerouzic commented on gnpy/core/network.py:
```diff
@@ -122,7 +139,7 @@ def select_edfa(raman_allowed, gain_target, power_target, equipment, uid):
                     -edfa.gain_min,
                 nf=edfa_nf(gain_target, edfa_variety, equipment)) \
                 for edfa_variety, edfa in edfa_dict.items()
-                if (edfa.allowed_for_design and not edfa.raman)]
```
Yes this part for rejected edfa.raman was and is still used ('and not edfa.raman' is still a condition for building this list)  because the filtering on suitable amp is not the same for Edfa and Raman amps.
Merging the future raman feature with autodesign will effectively have impacts. I have planned to work on that after july.
The condition on restriction is only here to enable amps from the restriction list and which are not allowed for the autodesign otherwise.
- allowed_for_design restricts the amps to use in general,
- restriction limits the amps to use for roadms only and overrides the allowed_for_design flag
this corresponds to real case where specific amp cards exist for some Roadms but not for others and not for ILA.

jktjkt replied:
It looks like I made a mistake interpretting the diff, sorry.

EstherLerouzic commented on gnpy/core/network.py:
```diff
@@ -222,6 +239,8 @@ def target_power(network, node, equipment): #get_fiber_dp
     if isinstance(node, Roadm):
         dp = 0
+    elif isinstance(node, Transceiver):
+        dp = -15
```
Current GNPy works for OSNR limited systems, meaning that it is assumed that the power level is guaranteed on receiver. Typical received power is around -15dBm. so we added this condition in this function. We saw this issue when testing the no-amp feature.

jktjkt replied:
Nice catch, then. However, right now the fix is buried in an unrelated commit and this comment is the only mention of why it is important. Can you please move this to a new pull request and explain in the commit message what you typed here?

I suspect that different transceivers/transponders have different threshold about a minimal RX power. Is that so, or am I wrong?

Also, it would be nice thing to cover this in an auto test -- that the transponders never end up with something like -20dBm at their ingress port.

EstherLerouzic replied:
May we postpone this ?
I would really like to have this PR merged and anyhow working in power limitation on transceiver would need a whole work, which is not a priority right now.
I would prefer to keep as is and add an issue to track this for enhancement and further work.

jktjkt replied:
I will create a small pull request doing just this change with a commit message clearly explaining why it is needed. This (#230) pull request can be then merged "on top of the by-then-updated `develop` branch, and the history will be less confusing for a future maintainer. Give me 5 minutes, OK? :)

EstherLerouzic replied:
OK !

jktjkt replied:
So this is more complex than I anticipated. If I simply add this on top current develop:
```diff
diff --git a/gnpy/core/network.py b/gnpy/core/network.py
index fec0fb0..bae429a 100644
--- a/gnpy/core/network.py
+++ b/gnpy/core/network.py
@@ -222,6 +222,8 @@ def target_power(network, node, equipment): #get_fiber_dp
 
     if isinstance(node, Roadm):
         dp = 0
+    elif isinstance(node, Transceiver):
+        dp = -15
 
     return dp
 
```
...then I'm getting a bunch of failures in `test_amplifier.py` such as this one:
```

gain = 15, nf_expected = 10
setup_edfa_fixed_gain = Edfa(uid='Edfa2', type_variety='test_fixed_gain', interpol_dgt=array([2.71452668, 2.71452668, 2.71452668, 2.71452668, ..., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]), pin_db=4.822712330395684, pout_db=5.066598484075778)
si = SpectralInformation(pref=Pref(p_span0=0.0, p_spani=0.0, neq_ch=19.822712330395685), carriers=[Channel(channel_number=1...=45000000000.0, roll_off=0.15, power=Power(signal=0.001, nonlinear_interference=0, amplified_spontaneous_emission=0))])

    @pytest.mark.parametrize("gain, nf_expected", [(15, 10), (20, 5), (25, 5)])
    def test_fixed_gain_nf(gain, nf_expected, setup_edfa_fixed_gain, si):
        """=> unitary test for fixed gain model Edfa._calc_nf() (and Edfa.interpol_params)"""
        edfa = setup_edfa_fixed_gain
        frequencies = array([c.frequency for c in si.carriers])
        pin = array([c.power.signal+c.power.nli+c.power.ase for c in si.carriers])
        pin = pin/db2lin(gain)
        baud_rates = array([c.baud_rate for c in si.carriers])
        edfa.operational.gain_target = gain
        pref = Pref(0, -gain, lin2db(len(frequencies)))
        edfa.interpol_params(frequencies, pin, baud_rates, pref)
    
>       assert pytest.approx(nf_expected, abs=0.01) == edfa.nf[0]
E       assert 10 ± 1.0e-02 == 25.0
E        +  where 10 ± 1.0e-02 = <function approx at 0x7f1be7017510>(10, abs=0.01)
E        +    where <function approx at 0x7f1be7017510> = pytest.approx

tests/test_amplifier.py:99: AssertionError
```
For me, that alone is a good reason to look at this change with even more suspicion than I had before :). So now I think we have two options:
- I just rubber-stamp this change as-is, based on your professional opinion that it's a good change. You clearly know more about this code than I do. At the same time, I do not like this option because I'm essentially merging code I do not understand.
- You help me understand what this thing (-15dBm minimal for transponders) really depends on. Does it depend on the `Fused` node? We separate this into a really smallest possible change, explain in the git log why it is happening, ensure that tests pass, and call it a day. Then we merge this (#230) change.

EstherLerouzic replied:
The error in test files is normal: reducing the target power has an big impact on the NF of the last edfa in tests where the trx is directly connected to an edfa. In order to keep performance unchanged and have a more realistic setup, an attenuator has to be put in fromt of the receiver. This is what I just pushed in PR #244 : the extraction from current PR of the changes related to this -15 dB + an attenuator added in the edfa_example_network.json
Hope this is clearer now

EstherLerouzic commented on gnpy/core/equipment.py:
```diff
@@ -364,6 +368,23 @@ def update_dual_stage(equipment):
+                pass
+        except KeyError:
+            print(
+                f'\x1b[1;31;40m'\
```
I understand your concern, except that it's more an "EquipmentError" than a "TopologyError" since it is just checking that the restriction in equipment_config json is valid.
I prepare this change

jktjkt commented:
And just FYI, it would be a bit easier for me if the work is split into small commits, each doing only one thing. This commit 038fe1835c3c5783e0adf2b272c5975a371ccea0 includes both trivial whitespace changes **and** replacing `print`s by logging plus exceptions. If these were separate, I would be able to glance over the trivial whitespace fixes very quickly, and focus on the more important changes.

If this is not a big problem for you, please consider splitting your work and committing more often.

jktjkt commented on gnpy/core/utils.py:
```diff
@@ -199,3 +199,48 @@ def rrc(ffs, baud_rate, alpha):
+    """Raised when the json equipment is not formed correctly
+    """
+    import sys
+    sys.tracebacklimit = 0
```
I do not think that this is a nice coding pattern. Raising an exception should not result in a random callback into the `sys` module and trying to reconfigure backtrace behavior. Also, [this variable is no longer available in Python 3](https://bugs.python.org/issue12276).

If the intention is to suppress detailed error messages, then I suggest moving any UI-affecting bits into the CLI driver, i.e., `examples/transmission_main_example.py`. I'll create another pull request for these exceptions.

EstherLerouzic commented:
hi @jktjkt ,
Is there anything else I should do for this PR to be at the level needed to be merged in develop ?

jktjkt commented:
> Is there anything else I should do for this PR to be at the level needed to be merged in develop ?

I would like to have these two addressed before this gets merged:
- Let's use exceptions and the associated error handling added in #251. I think there are two possibilities:
  - backporting this series on top of recent `develop`
  - doing that via git merges and resolving conflicts

Either you or me can do either of these two, please tell me what your preference is.
- Depend on that -15 dBm target power for transceivers which is done in #244. But I do not want #244 to be merged right away because it is not covered by tests. Again, two options IMHO;
  - wait for me to get #244 tested and merged
  - you can remove these -15dBm@transponder_RX change from this PR

Which one is better from your point of view?

I apologize that #244 is not in yet, but I got stuck at #249 (help from anyone is appreciated there, I simply do not understand the EDFA model enough to make progress).

EstherLerouzic commented:
OK  ! thanks for the feedback !
I prefer to do the backporting this series on top of recent `develop` 
and I will remove the -15dBm@transponder_RX change from this PR

jktjkt commented:
Thanks for completing this. Now the git history of this PR contains stuff which undoes another stuff done within this pull request, and I do not think that it's worth it to preserve such a level of detail. The new feature in general is reasonably compact, so it can be merged in a single commit (or two, with the `utils/` change separated out). I'll force-push a cleanup shortly.

jktjkt commented on gnpy/core/equipment.py:
```diff
@@ -134,7 +138,8 @@ def from_json(cls, filename, **kwargs):
         type_variety = kwargs['type_variety']
-        type_def = kwargs.get('type_def', 'variable_gain') #default compatibility with older json eqpt files
+        type_def = kwargs.get('type_def', 'variable_gain')
+        #default compatibility with older json eqpt files
```
I think that it's more intuitive if these comments are listed at the same line, or *before* the line they refer to. I've cleaned this.

jktjkt commented on gnpy/core/network.py:
```diff
@@ -50,9 +51,21 @@ def network_from_json(json_data, equipment):
             extra_params = equipment[typ][variety]
-            el_config.setdefault('params', {}).update(extra_params.__dict__)
+            temp = el_config.setdefault('params', {})
+            # update native function ovrwrite existing values: use a home made function
```
I've removed this comment. The function that we're calling is now named in a way which explains what this is about.

jktjkt commented:
@EstherLerouzic , can you please check that the descriptions given in the commit at the tip of this branch and in the README reflect reality? I tried to add a summary of what this thing is about into the commit message, and I also updated the README part about ROADMs.

EstherLerouzic commented:
@jktjkt, I think the sentence "In case a per-degree exception is needed, add the `Fused` node to a
specific degree. When the `Fused` element is present, the restrictions no longer apply, and amplifiers can be placed manually." 
should rather be something like:
"If a per-degree exception is needed, the amplifier of this degree can be defined in the Eqpt sheet or in the network json input. If no booster amplifier should be placed on a degree: 

  * (in case of an excel input) the term `fused` should be used for the amplifier type in the eqpt sheet. This will disable the automatic placement of an amplifier on this degree by setting a 'Fused' (logical) node instead. 
  * (In case of a json input), simply insert a 'Fused' node on the Roadm degree output.
"
Thanks for the changes !

jktjkt commented:
@EstherLerouzic , could you please check the updated wording? I've verified that you already put a correct wording into the Excel documentation, so I adjusted the commit message and clarified the JSON format in the main `README.rst`.

EstherLerouzic commented:
@jktjkt , the new wording is not correct for the fused part with excel input
'If no booster amplifier should be placed on a degree, use the `Fused`
node in place of an amplifier.'
because there will be an edfa placed between the ROADM and the 'Fused' node. This is why I have introduced the 'fused' (small 'f' ) keyword in the Eqpt parsing.
the correct sentence should be something like :
"If no booster amplifier should be placed on a degree, use a `Fused`
node in place of an amplifier in case of json input, or use the 'fused' keyword  for the amplifier type in the eqpt sheet for excel input. This will disable the automatic placement of an amplifier on this degree."

The Change in README is OK for me

EstherLerouzic commented:
I have done the requested changes.

jktjkt commented on tests/test_roadm_restrictions.py:
```diff
@@ -0,0 +1,198 @@
+
+# mark node_uid amps as fused for testing purpose
+@pytest.mark.parametrize("node_uid", ['east edfa in Lannion_CAS to Stbrieuc'])
+def test_no_amp_feature(node_uid):
```
I think that this test is no longer needed; the condition it was designed to test is no longer true. On the other hand, `examples/meshTopologyExampleV2.xls` and `examples/meshTopologyExampleV2.json` now "should be synced". (I'll add a TODO for me to turn this into a test later.)

So I went looking into `tests/data/`, and found `tests/data/testTopology.xls` which was not updated for recent changes. It also appears to have deviated from its likely origin, the `examples/meshTopologyExampleV2.xls`. I went ahead and tried making the requested change, but I think I got stuck somewhere because just running `convert.py` on my modified XLS produces results which I think are wrong.

Can you please also adjust `tests/data/testTopology.xls` so that it includes this new `fused` "amplifier" in that short span, and remove this test?

EstherLerouzic replied:
Not sure I understand what you mean by "no more needed" : there is still the possibility that a user needs a roadm with no booster amp, which is what I test here. I have changed the test yesterday: it checks that having a fused element placed just after a roadm disables the automatic addition of an EDFA there. I still think it is needed (to be sure future changes on parsers and build_network do not have regression on this). What is missing is the correct interpretation of fused keyword in parser. I will add that.

jktjkt replied:
I would have gone with a different approach where the test contains a predefined JSON file with exactly one `Fused` node at a ROADM egress, but I like this approach as well, and perhaps it's better. Thanks.

EstherLerouzic commented:
@jktjkt ,
in my last commits I corrected a bug (sorry) + updated tests and rebased all this on the latest develop . Could you check if this is OK for you ?
Besides, all tests are running ok on my computer, but the PR is not passing tests via travis on github: any idea ? (I have found no detailed report on the cause of this failure, and don't know what to correct). Thanks a lot for your help !

EstherLerouzic commented:
I found the issue !
  docstring was not correct with the bug correction, and since it is tested with travis, the test fails
